### PR TITLE
Add Phi model

### LINKED ISSUE #167: Adding phi model
Hello all.

I have started to work on adding microsoft phi models to Transformers, but I got totally stuck with the rotary positional encoding and the self-attention. I cannot get my head around it. @chengchingwen  Would you have a little bit of time for help?

Thanks a lot.

Tomas

My code so far is 
```julia
using ..Layers
using ..Layers: CompositeEmbedding, SelfAttention
using ChainRulesCore
using Functors
using Static

using NeuralAttentionlib
using NeuralAttentionlib: WithScore


abstract type HGFPhiPreTrainedModel <: HGFPreTrainedModel end

struct HGFPhiModel{E, DEC} <: HGFPhiPreTrainedModel
    embed::E
    decoder::DEC
end
@functor HGFPhiModel

(model::HGFPhiModel)(nt::NamedTuple) = model.decoder(model.embed(nt))

for T in :[
    HGFPhiForCausalLM,
    # HGFPhiForSequenceClassification,
].args
    @eval begin
        @hgfdefmodel $T HGFPhiPreTrainedModel
    end
end

basemodelkey(::Type{<:HGFPhiPreTrainedModel}) = :model
isbasemodel(::Type{<:HGFPhiModel}) = true
isbasemodel(::Type{<:HGFPhiPreTrainedModel}) = false

get_model_type(::Val{:phi}) = (
    model = HGFPhiModel,
    forcausallm = HGFPhiForCausalLM,
)


function load_model(_type::Type{HGFPhiPreTrainedModel}, cfg, state_dict, prefix)
    embed = load_model(HGFLlamaModel, CompositeEmbedding, cfg, state_dict, prefix)
    decoder = load_model(_type, TransformerBlock, cfg, state_dict, prefix)
    return HGFPhiModel(embed, decoder)
end

function load_model(_type::Type{HGFPhiForCausalLM}, cfg, state_dict, prefix)
    model = load_model(HGFPhiPreTrainedModel, cfg, state_dict, joinname(prefix, "model"))
    if cfg[:tie_word_embeddings]
        embedding = model.embed.token.embeddings
    else
        vocab_size, dims, factor = cfg[:vocab_size], cfg[:hidden_size], Float32(cfg[:initializer_range])
        embedding = getweight(weight_init(vocab_size, dims, factor), Layers.Embed,
                              state_dict, joinname(prefix, "lm_head.weight"))
    end
    lmhead = Layers.EmbedDecoder(Layers.Embed(embedding))
    return HGFLlamaForCausalLM(model, Layers.Branch{(:logit,), (:hidden_state,)}(lmhead))
end

function load_model(_type::Type{<:HGFPhiForCausalLM}, ::Type{<:SelfAttention}, cfg, state_dict, prefix)
    head, dims = cfg[:num_attention_heads], cfg[:hidden_size]
    @assert dims % head == 0 "The hidden size is not a multiple of the number of attention heads."
    head_dims = div(dims, head)
    kv_head = something(cfg[:num_key_value_heads], head)
    grouped_attn = kv_head != head
    @assert head % kv_head == 0 "The number of query is not dividable by the number of key/value groups"
    return_score = cfg[:output_attentions]
    factor = Float32(cfg[:initializer_range])
    @assert isnothing(cfg[:rope_scaling]) "Scaling Rotary Embedding is not support yet"
    @show (head_dims, head, kv_head)
    q_weight = getweight(weight_init(dims, dims, factor), Array,
                         state_dict, joinname(prefix, "q_proj.weight"))
    q_bias = getweight(weight_init(dims, dims, factor), Array,
                         state_dict, joinname(prefix, "q_proj.bias"))
    k_weight = getweight(weight_init(dims, kv_head * head_dims, factor), Array,
                         state_dict, joinname(prefix, "k_proj.weight"))
    k_bias = getweight(weight_init(dims, kv_head * head_dims, factor), Array,
                         state_dict, joinname(prefix, "k_proj.bias"))
    v_weight = getweight(weight_init(dims, kv_head * head_dims, factor), Array,
                         state_dict, joinname(prefix, "v_proj.weight"))
    v_bias = getweight(weight_init(dims, kv_head * head_dims, factor), Array,
                         state_dict, joinname(prefix, "v_proj.bias"))
    o_weight = getweight(weight_init(dims, dims, factor), Array, state_dict, joinname(prefix, "dense.weight"))
    o_bias = getweight(weight_init(dims, dims), Array, state_dict, joinname(prefix, "dense.bias"))
    qkv_proj = Layers.Fork(Layers.Dense(q_weight, q_bias), Layers.Dense(k_weight, k_bias), Layers.Dense(v_weight, v_bias))
    o_proj = Layers.Dense(o_weight, o_bias)
    if grouped_attn
        op = CausalLlamaRoPEGroupedQueryAttenOp(head, kv_head)
    else
        op = CausalGPTNeoXRoPEMultiheadQKVAttenOp(head_dims, head)
    end
    return_score && (op = WithScore(op))
    return SelfAttention(op, qkv_proj, o_proj)
end


function load_model(::Type{<:HGFPhiPreTrainedModel}, ::Type{<:Layers.LayerNorm}, cfg, state_dict, prefix)
    dims = cfg[:hidden_size]
    ln_ϵ = Float32(cfg[:layer_norm_eps])
    ln_weight = getweight(one_init(dims), Array, state_dict, joinname(prefix, "weight"))
    ln_bias = getweight(zero_init(dims), Array, state_dict, joinname(prefix, "bias"))
    return Layers.LayerNorm(ln_weight, ln_bias, ln_ϵ)
end

function load_model(_type::Type{<:HGFPhiPreTrainedModel}, ::Type{<:TransformerBlock}, cfg, state_dict, prefix)
    n = cfg[:num_hidden_layers]
    collect_output = cfg[:output_attentions] || cfg[:output_hidden_states]
    blocks = []
    for i = 1:n
        lprefix = joinname(prefix, :layers, i-1)

        ln = load_model(HGFPhiPreTrainedModel, Layers.LayerNorm, cfg, state_dict, joinname(lprefix, "input_layernorm"))
        sa = load_model(HGFLlamaPreTrainedModel, SelfAttention, cfg, state_dict, joinname(lprefix, "self_attn"))
        ff = load_model(HGFLlamaPreTrainedModel, Layers.Chain{Tuple{Layers.Dense, Layers.Dense}}, cfg, state_dict, joinname(lprefix, "mlp"))
        sa = Layers.PreNormResidual(sa, 

        block = TransformerBlock(sa, ff)
        push!(blocks, block)
    end
    collect_f = collect_output ? Layers.collect_outputs : nothing
    trf = Transformer(Tuple(blocks), collect_f)
    final_ln = load_model(HGFPhiPreTrainedModel, Layers.LayerNorm, cfg, state_dict, joinname(prefix, "norm"))
    return Layers.Chain(trf, final_ln)
end
```

### PR DESCRIPTION
**Pull request by chengchingwen**
I only test `microsoft/phi-1` with the validation code. Should generally give correct result, but would be better if we test other pre-trained model.

cc: @pevnak
Close #167

### CODE REVIEW

**Comment by pevnak** [comment]
Thanks for the quick work. 

I do not think I am getting similar outputs, but may-be, I am doing something wrong. I have written a python code to obtain intermediate results and save them as
```python
import torch
import transformers
import math
torch.manual_seed(0)

model = transformers.AutoModelForCausalLM.from_pretrained('microsoft/phi-1', torch_dtype=torch.float32, trust_remote_code=True)
tokenizer = transformers.AutoTokenizer.from_pretrained('microsoft/phi-1', trust_remote_code=True)

s = 'Tell me something about Julia?'
inputs = tokenizer(s, return_tensors='pt', return_attention_mask=False)
torch.save(inputs, '/tmp/inputs.torch')
e = model.model.embed_tokens(inputs.input_ids)
torch.save(e, '/tmp/embedding.torch')

x = e
for i in range(len(model.model.layers)):
   l = model.model.layers[i]
   x = l(hidden_states=x)[0]
   torch.save(x, f"/tmp/hidden_states_{i}.torch")
```

and then I load the model in Julia and try to verify the intermediate results
```julia
using Transformers
using Transformers.HuggingFace
using TextEncodeBase
using Statistics
using Pickle

function load_torch_matrix(filename)
    x = Pickle.Torch.THload(filename)
    x = Matrix(transpose(x[1,:,:]))
end

model_type = model_name = "microsoft/phi-1"
textenc = Transformers.HuggingFace.load_tokenizer(model_name)
cfg = Transformers.HuggingFace.load_config(model_type)
model = Transformers.HuggingFace.load_model("microsoft/phi-1")

s = "Tell me something about Julia?"

layers = model.decoder.layers[1];
x = load_torch_matrix("/tmp/embedding.torch")
for (i,l) in enumerate(layers.blocks)
    x = l((;hidden_state = x)).hidden_state
    r = load_torch_matrix("/tmp/hidden_states_$(i-1).torch")
    println(i,": ",maximum(abs.(r .- x)))
end
```
and I obtain big differences
```
1: 4.5420475
2: 7.1327615
3: 86.075935
4: 188.93604
5: 274.75513
6: 323.44397
7: 348.9238
8: 355.41754
9: 358.76648
10: 358.0651
11: 357.5178
12: 354.0831
13: 350.16187
14: 342.29797
15: 334.9134
16: 327.589
17: 319.62167
18: 301.97784
19: 269.8337
20: 226.61732
21: 158.06534
22: 93.10416
23: 54.30915
24: 42.18445
```

Do you have an idea, what I did wrong?

Thanks for help,
Tomas

**Comment by chengchingwen** [comment]
You are not calling the python model correctly, the python implementation of `PhiDecoderLayer` require passing the correct attention mask for performing causal attention.

https://github.com/huggingface/transformers/blob/03cc17775b961d16cc4d0d7ab0c8487120d0b708/src/transformers/models/phi/modeling_phi.py#L900

Generally, when using huggingface transformers to get the per layers hidden state, it is best to use the `return_hidden_states=True` in the config.

**Comment by pevnak** [comment]
Great. So I have modified my scripts for comparison as follows.
```python
# I let julia handle conda and stuff there
# add PythonCall CondaPkg
# ] conda add pytorch
# ] conda add math
# /private/tmp/.CondaPkg/env/bin/pip install git+https://github.com/huggingface/transformers
# /private/tmp/.CondaPkg/env/bin/pip install ipython
import torch
import transformers
import math
model = transformers.AutoModelForCausalLM.from_pretrained('microsoft/phi-1', torch_dtype=torch.float32, trust_remote_code=True)
tokenizer = transformers.AutoTokenizer.from_pretrained('microsoft/phi-1', trust_remote_code=True)

s = 'Tell me something about Julia?'
inputs = tokenizer(s, return_tensors='pt', return_attention_mask=True)
attention_mask = inputs.attention_mask
torch.save(inputs.input_ids, '/tmp/inputs.torch')
e = model.model.embed_tokens(inputs.input_ids)
torch.save(e, '/tmp/embedding.torch')

outputs = model.model(**inputs, output_hidden_states=True)
for i in range(len(outputs.hidden_states)):
   x = outputs.hidden_states[i]
   torch.save(x, f"/tmp/hidden_states_{i}.torch")
```
and julia
```julia
using Transformers
using Transformers.HuggingFace
using TextEncodeBase
using Statistics
using Pickle

function load_torch_matrix(filename)
    x = Pickle.Torch.THload(filename)
    x = Matrix(transpose(x[1,:,:]))
end

model_name = "microsoft/phi-1"
textenc = Transformers.HuggingFace.load_tokenizer(model_name)
model = Transformers.HuggingFace.load_model(model_name)

s = "Tell me something about Julia?"

layers = model.decoder.layers[1];
x = load_torch_matrix("/tmp/hidden_states_0.torch")
for (i,l) in enumerate(layers.blocks)
    x = l((;hidden_state = x)).hidden_state
    if i == 24 
        x = model.decoder.layers[2](x)
    end
    r = load_torch_matrix("/tmp/hidden_states_$(i).torch")
    println(i,": linfty: ",maximum(abs.(r .- x)), " l1: ", mean(abs.(r .- x)))
    x = r # this is in order not to accumulate error
end
```

which gives following output
```
1: linfty: 0.039717197 l1: 0.0053815306
2: linfty: 6.1035156e-5 l1: 3.9946585e-6
3: linfty: 0.0002670288 l1: 3.3049548e-6
4: linfty: 3.7670135e-5 l1: 1.9887095e-6
5: linfty: 3.0517578e-5 l1: 1.0879072e-6
6: linfty: 2.336502e-5 l1: 1.0066275e-6
7: linfty: 3.0517578e-5 l1: 8.849602e-7
8: linfty: 1.8119812e-5 l1: 8.8099995e-7
9: linfty: 3.0517578e-5 l1: 6.895537e-7
10: linfty: 1.9073486e-5 l1: 8.302106e-7
11: linfty: 4.7683716e-6 l1: 5.8953475e-7
12: linfty: 7.6293945e-6 l1: 6.3155375e-7
13: linfty: 8.583069e-6 l1: 5.053671e-7
14: linfty: 3.8146973e-6 l1: 5.0470317e-7
15: linfty: 3.0517578e-5 l1: 6.371617e-7
16: linfty: 3.8146973e-6 l1: 4.999047e-7
17: linfty: 3.8146973e-6 l1: 5.8280847e-7
18: linfty: 3.0517578e-5 l1: 6.3953786e-7
19: linfty: 1.5258789e-5 l1: 7.439315e-7
20: linfty: 1.5258789e-5 l1: 8.079199e-7
21: linfty: 1.5258789e-5 l1: 6.644962e-7
22: linfty: 1.5258789e-5 l1: 7.874987e-7
23: linfty: 1.5258789e-5 l1: 8.248753e-7
24: linfty: 1.66893e-6 l1: 2.4516575e-7
```
which is OK. The discrepancy I think is caused by a different implementation of LayerNorm between Torch and NeuralAttentionLib.
NeuralAttentionLib implements the LayerNorm as 
```julia
function _normalize(inN::T, ϵ::T, x::T, mean_M2_::Tuple{Int32, T, T}) where T
    _, μ, M2 = mean_M2_
    v = M2 * inN
    σ₀ = sqrt(v .+ ϵ)
    σ = max(σ₀, ϵ)
    return (x - μ) / σ
end
```
whereas torch implements it as 
```julia
function _normalize(inN::T, ϵ::T, x::T, mean_M2_::Tuple{Int32, T, T}) where T
    _, μ, M2 = mean_M2_
    v = M2 * inN
    σ = sqrt(v .+ ϵ)
    return (x - μ) / σ
end
```
When I change the implementation to the latter, the differences are 
```julia
1: linfty: 4.172325e-6 l1: 4.0366422e-7
2: linfty: 2.861023e-6 l1: 3.1565227e-7
3: linfty: 7.6293945e-5 l1: 6.0604117e-7
4: linfty: 6.1035156e-5 l1: 5.625631e-7
5: linfty: 4.5776367e-5 l1: 3.6759593e-7
6: linfty: 9.536743e-6 l1: 4.0165318e-7
7: linfty: 3.0517578e-5 l1: 4.8193243e-7
8: linfty: 5.2452087e-6 l1: 4.3385535e-7
9: linfty: 3.0517578e-5 l1: 4.046833e-7
10: linfty: 1.1444092e-5 l1: 5.818394e-7
11: linfty: 5.722046e-6 l1: 4.9513864e-7
12: linfty: 4.7683716e-6 l1: 5.0985244e-7
13: linfty: 8.106232e-6 l1: 4.821841e-7
14: linfty: 3.8146973e-6 l1: 4.850729e-7
15: linfty: 3.0517578e-5 l1: 5.4865603e-7
16: linfty: 3.8146973e-6 l1: 4.811197e-7
17: linfty: 4.7683716e-6 l1: 5.971572e-7
18: linfty: 3.0517578e-5 l1: 6.217172e-7
19: linfty: 1.5258789e-5 l1: 6.079869e-7
20: linfty: 1.5258789e-5 l1: 6.9474845e-7
21: linfty: 2.2888184e-5 l1: 6.9902245e-7
22: linfty: 1.5258789e-5 l1: 6.93403e-7
23: linfty: 1.1444092e-5 l1: 8.5710843e-7
24: linfty: 1.9073486e-6 l1: 2.1992787e-7
```
which is much closer to the python. I think that the current implementation in Julia might still produce `nan` when the variance is zero, because the gradient of standard deviation is `1/(2*sqrt(v))` which will produce nan when `v` is equal zero. 
This is a question if we want to be equal to pytorch (I think that in this lib likely yes) or be more mathematically correct, in which case the implementation should be as follows to get the best of both worlds (be mathematically correct normalization everywhere except `[0, ϵ]`). 
```julia
function _normalize(inN::T, ϵ::T, x::T, mean_M2_::Tuple{Int32, T, T}) where T
    _, μ, M2 = mean_M2_
    v₀ = M2 * inN
     v = max(v₀, ϵ)
    σ = sqrt(v)
    return (x - μ) / σ
end
```
 
 Anyway, I do not think this is a big deal. But people can be scared when results do not match exactly.
 
 I really appreciate your help Peter. I am all the time amazed by your work and level of support.
 
 I have tried to load phi-1.5, but I got an error
 ```julia
 ERROR: Distant resource does not have an ETag, we won't be able to reliably ensure reproducibility.
```
For phi-2, the problem is that it saves weights in two `safetensor` files and does not contain picles.

**Comment by chengchingwen** [comment]
> . I think that the current implementation in Julia might still produce nan when the variance is zero, because the gradient of standard deviation is 1/(2*sqrt(v)) which will produce nan when v is equal zero.

This should not happened. The layer norm in nalib has a hand-written gradient which should handle that case already. That change is for handling some cases that the precision is too small to handle the variance correctly (e.g. `sqrt(Float16(1e-9)) == 0`).

> I have tried to load phi-1.5, but I got an error

Which pre-trained model are you testing? `HuggingFace.hgf"microsoft/phi-1_5"` seems to work correctly on my machine.

> For phi-2, the problem is that it saves weights in two safetensor files and does not contain picles.

We need either make PR to phi-2 with the weights converted to pickles or see if we can directly support safetensor in Julia.

**Comment by pevnak** [comment]
Hi Peter,

I made a typo in specification of the model. I have used dot, `phi_1.5` instead of underscore `phi_1_5`, which explains the problem.

I have sketched a first try on parser of safetensors and posted it here https://discourse.julialang.org/t/parser-for-safetensors/109366 It seems to me to do the job by naked eye, but I have tested it just on one file. I hope to try the phi_1_5 model tomorrow.

Handling the layer normalization is neat!!!

**Comment by pevnak** [comment]
This is superceded by #169
